### PR TITLE
add tf.mask_fill()

### PR DESCRIPTION
**Pull request by aavishkarmishra**
## Added new api  `tf.mask_fill()` according to issue #41617 

> This api will work similar to [Pytorch's masked_fill_](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.masked_fill_)

### Example:
`>>>tensor = tf.constant([1, 2, 3])`
`>>>mask = tf.constant([True, False, True])`
`>>>value = 5`
`>>>tf.mask_fill(tensor, mask, value)`
`<tf.Tensor: shape=(3,), dtype=int32, numpy=array([5, 2, 5], dtype=int32)>`

### CODE REVIEW

File: tensorflow/python/ops/array_ops.py [hunk 1]
```diff
@@ -5871,3 +5871,55 @@ def repeat(input, repeats, axis=None, name=None):  # pylint: disable=redefined-b
...
+  corresponding index have `True` or `1` entry in `mask` array.
+
+  Examples:
+
+  ```python
+  tensor = [0, 1, 2, 3]
+  mask = np.array([True, False, True, False])
+  value = 5
+  tf.mask_fill(tensor, mask, value)  # [5, 1, 5, 3]
+  ```
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/array_ops.py:5890]: Let's write this example as a python docstring. See https://www.tensorflow.org/community/contribute/docs_ref#testable_docstrings

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/array_ops.py [hunk 2]
```diff
@@ -5871,3 +5871,55 @@ def repeat(input, repeats, axis=None, name=None):  # pylint: disable=redefined-b
     input = reshape(input, [-1])
     axis = 0
   return repeat_with_axis(input, repeats, axis, name)
+
+@tf_export(v1=["mask_fill"])
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/array_ops.py:5875]:
```suggestion
@tf_export("mask_fill")
```

The `v1=[...` part makes the API available only as part of `tf.compat.v1` and TF1.15

File: tensorflow/python/ops/array_ops.py [hunk 3]
```diff
@@ -5871,3 +5871,55 @@ def repeat(input, repeats, axis=None, name=None):  # pylint: disable=redefined-b
...
+  corresponding index have `True` or `1` entry in `mask` array.
+
+  Examples:
+
+  ```python
+  >>>tensor = [0, 1, 2, 3]
+  >>>mask = np.array([True, False, True, False])
+  >>>value = 5
+  >>>tf.mask_fill(tensor, mask, value)  # [5, 1, 5, 3]
+  ```
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/array_ops.py:5890]:
```suggestion
  >>> tensor = tf.constant([0, 1, 2, 3])
  >>> mask = np.array([True, False, True, False])
  >>> tf.mask_fill(tensor, mask, 5)
  [5, 1, 5, 3]
```

(something like that)

File: tensorflow/python/ops/array_ops.py (hunk 1 above)
Comment (aavishkarmishra) [reply tensorflow/python/ops/array_ops.py]: I have a doubt that will `tf.mask_fill()` will work correctly as the name of function is `mask_fill`

File: tensorflow/python/ops/array_ops.py [hunk 4]
```diff
@@ -5871,3 +5871,55 @@ def repeat(input, repeats, axis=None, name=None):  # pylint: disable=redefined-b
...
+    1-dimensional tensor updatd by entries in `tensor` corresponding
+    to `True` values in `mask` to the `value`.
+
+  Raises:
+    ValueError:  If shapes do not conform.
+  """
+
+  def mask__fill_1d(tensor, mask, value):
+    """Mask tensor along dimension 0 with a 1-D mask."""
+    return tensor * (1 - tf.cast(mask, tf.int32)) + value * tf.cast(mask, tf.int32)
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/array_ops.py:5909]: Why the cast to `tf.int32`? Shouldnt we instead cast to the same type as `tensor`?

File: tensorflow/python/ops/array_ops.py [hunk 5]
```diff
@@ -5871,3 +5871,55 @@ def repeat(input, repeats, axis=None, name=None):  # pylint: disable=redefined-b
...
+
+    shape_mask = mask.get_shape()
+    ndims_mask = shape_mask.ndims
+    shape_tensor = tensor.get_shape()
+    if ndims_mask == 0:
+      raise ValueError("mask cannot be scalar.")
+    if ndims_mask is None:
+      raise ValueError(
+          "Number of mask dimensions must be specified, even if some dimensions"
+          " are None.  E.g. shape=[None] is ok, but shape=None is not.")
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/array_ops.py:5923]: Should we have a check that the mask is a boolean array?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/array_ops.py (hunk 5 above)
Comment (aavishkarmishra) [reply tensorflow/python/ops/array_ops.py:5923]: In agrs section it is mentioned that mask should be a boolean array, but if you want then I can throw error if mask is not boolean array .

Comment (aavishkarmishra) [reply tensorflow/python/ops/array_ops.py:5927]: I have added a check, please check that .

**Comment by alextp** [comment]
Isn't tf.mask_fill(tensor, mask, value) equivalent tf.where(mask, value, tensor)? (and the implementation of tf.where is better than the multiply-by-mask one.

**Pull request closed without merging**